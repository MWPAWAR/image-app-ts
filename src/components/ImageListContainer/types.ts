export interface ImageListContainerProps {
  activeTab: string
}
