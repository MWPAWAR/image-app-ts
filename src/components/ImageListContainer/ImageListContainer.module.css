.mainContainer {
  display: flex;
}

.heading {
  margin: 30px 15px 15px 15px;
}

.listContainerFullWidth {
  background-color: #fafafa;
  width: 100%;
}

.listContainerSplitWidth {
  background-color: #fafafa;
  width: 100%;
}

.sheetContainer {
  background-color: #fff;
  border-top: outset 2px;
  width: 400px !important;
  overflow-y: scroll;
}
