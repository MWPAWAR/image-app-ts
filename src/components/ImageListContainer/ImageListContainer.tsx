import { useState } from 'react'
import Drawer from 'react-modern-drawer'
import 'react-modern-drawer/dist/index.css'
import { useAppSelector } from '../customHooks/hooks'
import useFetchImages from '../customHooks/useFetchImages'
import ImageList from '../ImageList/ImageList'
import ImageSheet from '../ImageSheet/ImageSheet'
import withApiLoaderOrFailure from '../hocs/with-api-loader-or-failure'
import TabList from '../TabList/TabList'
import { selectImageById } from '../../store/imageUiState/selectors'
import {
  recentlyAddedImages,
  favoriteImages,
} from '../../store/images/selectors'
import imageListContainerStyles from './ImageListContainer.module.css'
import { ImageListContainerProps } from './types'
import { getString } from '../../i18n'
import { ImageListTypes, ImageTypes } from '../types'

const ImageListContainer = ({ activeTab }: ImageListContainerProps) => {
  useFetchImages()

  const selectedImage: ImageTypes | undefined = useAppSelector(selectImageById)
  const recentlyAddedImageList: ImageListTypes =
    useAppSelector(recentlyAddedImages)
  const favoriteImageList: ImageListTypes = useAppSelector(favoriteImages)
  const imageList: ImageListTypes =
    activeTab === 'recentlyAdded' ? recentlyAddedImageList : favoriteImageList

  const [isOpen, setIsOpen] = useState<boolean>(!!selectedImage)
  const toggleDrawer = () => {
    setIsOpen((prevState) => !prevState)
  }

  return (
    <div className={imageListContainerStyles.mainContainer}>
      <div
        className={
          selectedImage
            ? imageListContainerStyles.listContainerSplitWidth
            : imageListContainerStyles.listContainerFullWidth
        }
      >
        <h1 className={imageListContainerStyles.heading}>
          {getString('imageListContainer.photosHeading')}
        </h1>
        <TabList />
        <ImageList imageList={imageList} toggleDrawer={toggleDrawer} />
      </div>
      <Drawer
        open={isOpen}
        onClose={toggleDrawer}
        direction="right"
        className={imageListContainerStyles.sheetContainer}
      >
        <ImageSheet activeTab={activeTab} toggleDrawer={toggleDrawer} />
      </Drawer>
    </div>
  )
}

export default withApiLoaderOrFailure(ImageListContainer)
