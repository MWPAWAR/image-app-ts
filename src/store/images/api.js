export default () => fetch('https://agencyanalytics-api.vercel.app/images.json')
  .then((response) => {
    if (!response.ok) throw new Error('Failure')
    return response.json();
  });