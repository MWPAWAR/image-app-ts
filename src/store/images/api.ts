import { ImagesFetchApiResponse } from './types'

export default (): Promise<ImagesFetchApiResponse> =>
  fetch('https://agencyanalytics-api.vercel.app/images.json').then(
    (response): Promise<ImagesFetchApiResponse> => {
      if (!response.ok) throw new Error('Failure')
      return response.json()
    }
  )
