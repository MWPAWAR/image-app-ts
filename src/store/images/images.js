import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import getImages from './api';

export const fetchImages = createAsyncThunk('images/fetchImages', async () => {
  const response = await getImages();
  return response;
});

const imagesSlice = createSlice({
  name: 'images',
  initialState: { status: 'idle', images: [] },
  reducers: {
    toggleFavoriteImg(state, action) {
      const index = state.images.findIndex((image) => image.id === action.payload);
      if (typeof index !== 'undefined') {
        state.images[index].favorited = !state.images[index].favorited;
      }
    },
    deleteImg(state, action) {
      const index = state.images.findIndex((image) => image.id === action.payload);
      if (typeof index !== 'undefined') {
        state.images.splice(index, 1);
      }
    }
  },
  extraReducers(builder) {
    builder
      .addCase(fetchImages.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchImages.fulfilled, (state, action) => {
        state.status = 'success';
        state.images = action.payload;
      })
      .addCase(fetchImages.rejected, (state) => {
        state.error = 'error';
      })
  }
});

export const { toggleFavoriteImg, deleteImg } = imagesSlice.actions;
export default imagesSlice.reducer;
