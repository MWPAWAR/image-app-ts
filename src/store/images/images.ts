import { createSlice, createAsyncThunk } from '@reduxjs/toolkit'
import { ImageListTypes } from '../../components/types'
import getImages from './api'
import { API_FETCH_STATUS } from '../../constants'
import {
  ImagesState,
  ImagesExtraReducerStateAction,
  ImagesReducerStateAction,
  ImagesFetchApiResponse,
} from './types'

export const fetchImages = createAsyncThunk<ImagesFetchApiResponse>(
  'images/fetchImages',
  async () => {
    const response: ImageListTypes = await getImages()
    return response
  }
)

const initialState: ImagesState = {
  status: API_FETCH_STATUS.IDLE,
  images: [],
}

const imagesSlice = createSlice({
  name: 'images',
  initialState,
  reducers: {
    toggleFavoriteImg(
      state: ImagesState,
      action: ImagesReducerStateAction
    ): void {
      const index: number | undefined = state.images.findIndex(
        (image) => image.id === action.payload.id
      )
      if (typeof index !== 'undefined') {
        state.images[index].favorited = !state.images[index].favorited
      }
    },
    deleteImg(state: ImagesState, action: ImagesReducerStateAction): void {
      const index: number | undefined = state.images.findIndex(
        (image) => image.id === action.payload.id
      )
      if (typeof index !== 'undefined') {
        state.images.splice(index, 1)
      }
    },
  },
  extraReducers(builder) {
    builder
      .addCase(fetchImages.pending, (state: ImagesState) => {
        state.status = API_FETCH_STATUS.LOADING
      })
      .addCase(
        fetchImages.fulfilled,
        (state: ImagesState, action: ImagesExtraReducerStateAction) => {
          state.status = API_FETCH_STATUS.SUCCESS
          state.images = action.payload
        }
      )
      .addCase(fetchImages.rejected, (state: ImagesState) => {
        state.status = API_FETCH_STATUS.ERROR
      })
  },
})

export const { toggleFavoriteImg, deleteImg } = imagesSlice.actions
export default imagesSlice.reducer
