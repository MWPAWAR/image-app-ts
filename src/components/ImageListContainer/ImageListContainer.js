import { Fragment } from 'react';
import { useSelector } from 'react-redux';
import useFetchImages from '../../customHooks/useFetchImages';
import ImageList from '../ImageList/ImageList';
import ImageSheet from '../ImageSheet/ImageSheet';
import withApiLoaderOrFailure from '../hocs/with-api-loader-or-failure';
import TabList from '../TabList/TabList';
import { selectImageById } from '../../store/imageUiState/selectors';
import { recentlyAddedImages, favoriteImages } from '../../store/images/selectors';
import imageListContainerStyles from './ImageListContainer.module.css';

const ImageListContainer = ({ activeTab }) => {
  useFetchImages();

  const selectedImage = useSelector(selectImageById);
  const recentlyAddedImageList = useSelector(recentlyAddedImages);
  const favoriteImageList = useSelector(favoriteImages);
  const imageList = activeTab === 'recentlyAdded'
    ? recentlyAddedImageList
    : favoriteImageList;

  return (
    <div className={imageListContainerStyles.mainContainer}>
      <div className={selectedImage ? imageListContainerStyles.listContainerSplitWidth : imageListContainerStyles.listContainerFullWidth}>
        <h1 className={imageListContainerStyles.heading}>Photos</h1>
        <TabList />
        <ImageList imageList={imageList} />
      </div>
      {selectedImage ? (
        <div className={imageListContainerStyles.sheetContainer}>
          <ImageSheet activeTab={activeTab} />
        </div>
      ) : <Fragment />}
    </div>
  );
};

export default withApiLoaderOrFailure(ImageListContainer);
