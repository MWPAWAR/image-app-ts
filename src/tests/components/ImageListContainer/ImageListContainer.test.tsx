import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import { imageMock1 } from '../../mocks/image'
import { images } from '../../mocks/images'
import ImageListContainer from '../../../components/ImageListContainer/ImageListContainer'

describe('ImageListContainer', () => {
  const mockStore = configureStore([])
  const store = mockStore({
    images: {
      images,
    },
    imageUiState: {
      selectedImg: '',
    },
  })
  jest.spyOn(store, 'dispatch')

  it('ImageListContainer matches snapshot for Recently added tab', () => {
    const Component = (
      <MemoryRouter>
        <Provider store={store}>
          <ImageListContainer activeTab="recentlyAdded" />
        </Provider>
      </MemoryRouter>
    )
    const tree = renderer.create(Component)
    expect(tree).toMatchSnapshot()
  })

  it('ImageListContainer matches snapshot for Favorited tab', () => {
    const Component = (
      <MemoryRouter>
        <Provider store={store}>
          <ImageListContainer activeTab="favorited" />
        </Provider>
      </MemoryRouter>
    )
    const tree = renderer.create(Component)
    expect(tree).toMatchSnapshot()
  })

  it('ImageListContainer matches snapshot if user have selected an image', () => {
    const mockStore = configureStore([])
    const store = mockStore({
      images: {
        images,
      },
      imageUiState: {
        selectedImg: imageMock1.id,
      },
    })

    const Component = (
      <MemoryRouter>
        <Provider store={store}>
          <ImageListContainer activeTab="recentlyAdded" />
        </Provider>
      </MemoryRouter>
    )
    const tree = renderer.create(Component)
    expect(tree).toMatchSnapshot()
  })
})
