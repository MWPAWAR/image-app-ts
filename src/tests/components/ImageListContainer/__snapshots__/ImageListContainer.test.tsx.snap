// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageListContainer ImageListContainer matches snapshot for Favorited tab 1`] = `
<div
  className="mainContainer"
>
  <div
    className="listContainerFullWidth"
  >
    <h1
      className="heading"
    >
      Photos
    </h1>
    <div
      className="tabList"
    >
      <div
        className="tabListItem"
        onClick={[Function]}
      >
        Recently added
      </div>
      <div
        className="tabListItem"
        onClick={[Function]}
      >
        Favorited
      </div>
    </div>
    <div
      className="imageList"
    >
      <div
        className="imageListItem"
        data-testid="image-list-item"
        onClick={[Function]}
      >
        <img
          alt="tennessee_female_rubber.jpg"
          className="imageListItemImg"
          src="https://agencyanalytics-api.vercel.app/images/0.jpg"
          tabIndex={0}
        />
        <div
          className="imageListItemName"
        >
          tennessee_female_rubber.jpg
        </div>
        <div
          className="imageListItemSize"
        >
          4.6 MB
        </div>
      </div>
    </div>
  </div>
  <div
    className="EZDrawer"
    id="EZDrawer6svab"
  >
    <input
      checked={false}
      className="EZDrawer__checkbox"
      id="EZDrawer__checkbox6svab"
      onChange={[Function]}
      type="checkbox"
    />
    <nav
      className="EZDrawer__container sheetContainer"
      id="EZDrawer__container6svab"
      role="navigation"
      style={
        Object {
          "height": "100vh",
          "right": 0,
          "top": 0,
          "transform": "translate3d(100%, 0, 0)",
          "transitionDuration": "500ms",
          "width": 250,
          "zIndex": 101,
        }
      }
    />
    <label
      className="EZDrawer__overlay"
      htmlFor="EZDrawer__checkbox6svab"
      id="EZDrawer__overlay6svab"
      style={
        Object {
          "backgroundColor": "#000",
          "opacity": "0.4",
          "zIndex": 100,
        }
      }
    />
  </div>
</div>
`;

exports[`ImageListContainer ImageListContainer matches snapshot for Recently added tab 1`] = `
<div
  className="mainContainer"
>
  <div
    className="listContainerFullWidth"
  >
    <h1
      className="heading"
    >
      Photos
    </h1>
    <div
      className="tabList"
    >
      <div
        className="tabListItem"
        onClick={[Function]}
      >
        Recently added
      </div>
      <div
        className="tabListItem"
        onClick={[Function]}
      >
        Favorited
      </div>
    </div>
    <div
      className="imageList"
    >
      <div
        className="imageListItem"
        data-testid="image-list-item"
        onClick={[Function]}
      >
        <img
          alt="virginia.jpg"
          className="imageListItemImg"
          src="https://agencyanalytics-api.vercel.app/images/1.jpg"
          tabIndex={0}
        />
        <div
          className="imageListItemName"
        >
          virginia.jpg
        </div>
        <div
          className="imageListItemSize"
        >
          4.1 MB
        </div>
      </div>
      <div
        className="imageListItem"
        data-testid="image-list-item"
        onClick={[Function]}
      >
        <img
          alt="tennessee_female_rubber.jpg"
          className="imageListItemImg"
          src="https://agencyanalytics-api.vercel.app/images/0.jpg"
          tabIndex={0}
        />
        <div
          className="imageListItemName"
        >
          tennessee_female_rubber.jpg
        </div>
        <div
          className="imageListItemSize"
        >
          4.6 MB
        </div>
      </div>
    </div>
  </div>
  <div
    className="EZDrawer"
    id="EZDrawery3fo5h"
  >
    <input
      checked={false}
      className="EZDrawer__checkbox"
      id="EZDrawer__checkboxy3fo5h"
      onChange={[Function]}
      type="checkbox"
    />
    <nav
      className="EZDrawer__container sheetContainer"
      id="EZDrawer__containery3fo5h"
      role="navigation"
      style={
        Object {
          "height": "100vh",
          "right": 0,
          "top": 0,
          "transform": "translate3d(100%, 0, 0)",
          "transitionDuration": "500ms",
          "width": 250,
          "zIndex": 101,
        }
      }
    />
    <label
      className="EZDrawer__overlay"
      htmlFor="EZDrawer__checkboxy3fo5h"
      id="EZDrawer__overlayy3fo5h"
      style={
        Object {
          "backgroundColor": "#000",
          "opacity": "0.4",
          "zIndex": 100,
        }
      }
    />
  </div>
</div>
`;

exports[`ImageListContainer ImageListContainer matches snapshot if user have selected an image 1`] = `
<div
  className="mainContainer"
>
  <div
    className="listContainerSplitWidth"
  >
    <h1
      className="heading"
    >
      Photos
    </h1>
    <div
      className="tabList"
    >
      <div
        className="tabListItem"
        onClick={[Function]}
      >
        Recently added
      </div>
      <div
        className="tabListItem"
        onClick={[Function]}
      >
        Favorited
      </div>
    </div>
    <div
      className="imageList"
    >
      <div
        className="imageListItem"
        data-testid="image-list-item"
        onClick={[Function]}
      >
        <img
          alt="virginia.jpg"
          className="imageListItemImg"
          src="https://agencyanalytics-api.vercel.app/images/1.jpg"
          tabIndex={0}
        />
        <div
          className="imageListItemName"
        >
          virginia.jpg
        </div>
        <div
          className="imageListItemSize"
        >
          4.1 MB
        </div>
      </div>
      <div
        className="imageListItem"
        data-testid="image-list-item"
        onClick={[Function]}
      >
        <img
          alt="tennessee_female_rubber.jpg"
          className="imageListItemImg"
          src="https://agencyanalytics-api.vercel.app/images/0.jpg"
          tabIndex={0}
        />
        <div
          className="imageListItemName"
        >
          tennessee_female_rubber.jpg
        </div>
        <div
          className="imageListItemSize"
        >
          4.6 MB
        </div>
      </div>
    </div>
  </div>
  <div
    className="EZDrawer"
    id="EZDrawerpgzobk"
  >
    <input
      checked={true}
      className="EZDrawer__checkbox"
      id="EZDrawer__checkboxpgzobk"
      onChange={[Function]}
      type="checkbox"
    />
    <nav
      className="EZDrawer__container sheetContainer"
      id="EZDrawer__containerpgzobk"
      role="navigation"
      style={
        Object {
          "height": "100vh",
          "right": 0,
          "top": 0,
          "transform": "translate3d(100%, 0, 0)",
          "transitionDuration": "500ms",
          "width": 250,
          "zIndex": 101,
        }
      }
    >
      <div
        className="imageSheet"
      >
        <div>
          <img
            alt="tennessee_female_rubber.jpg"
            className="image"
            src="https://agencyanalytics-api.vercel.app/images/0.jpg"
          />
          <div
            className="imgNameAndFavtIconContainer"
          >
            <div
              className="imageName"
            >
              tennessee_female_rubber.jpg
            </div>
            <div
              className="favtIcon favtIconFilled"
              data-testid="fav-icon"
              onClick={[Function]}
            />
          </div>
          <div
            className="imageSize"
          >
            4.6 MB
          </div>
        </div>
        <div
          className="imgDetails"
        >
          <div
            className="informationLabel"
          >
            Information
          </div>
          <div
            className="row"
          >
            <div
              className="label"
            >
              Uploaded by
            </div>
            <div
              className="value"
            >
              Ms. Jimmie Cole
            </div>
          </div>
          <div
            className="row"
          >
            <div
              className="label"
            >
              Created
            </div>
            <div
              className="value"
            >
              July 15, 2017
            </div>
          </div>
          <div
            className="row"
          >
            <div
              className="label"
            >
              Last modified
            </div>
            <div
              className="value"
            >
              December 16, 2022
            </div>
          </div>
          <div
            className="row"
          >
            <div
              className="label"
            >
              Dimensions
            </div>
            <div
              className="value"
            >
              4800 X 3200
            </div>
          </div>
          <div
            className="row"
          >
            <div
              className="label"
            >
              Resolutions
            </div>
            <div
              className="value"
            >
              72 X 72
            </div>
          </div>
        </div>
        <div
          className="imgDescription"
        >
          <div
            className="label"
          >
            Description
          </div>
          <div
            className="description"
          >
            Laboriosam eligendi inventore officia nemo. Quisquam explicabo voluptatem. Illo laborum facilis.
          </div>
        </div>
        <div
          className="imageSheetFooter"
        >
          <div
            className="deleteBtn"
          >
            <button
              className="button"
              onClick={[Function]}
            >
              Delete
            </button>
          </div>
          <div
            className="backBtn"
          >
            <button
              className="button"
              onClick={[Function]}
            >
              Back
            </button>
          </div>
        </div>
      </div>
    </nav>
    <label
      className="EZDrawer__overlay"
      htmlFor="EZDrawer__checkboxpgzobk"
      id="EZDrawer__overlaypgzobk"
      style={
        Object {
          "backgroundColor": "#000",
          "opacity": "0.4",
          "zIndex": 100,
        }
      }
    />
  </div>
</div>
`;
